internal/plugin/connectors/http/generic/README.md

```diff
@@ -0,0 +1,201 @@
+digits.  If this rule is violated at runtime, Secretless will log an
+appropriate error.
+
+### Limitations
+
+Some HTTP APIs require more involved authentication.  For example, they might
+require you to read the HTTP body content and create a hashed signature of it
+using a secret key.
+
+For now, the Generic HTTP connector does not support use cases like these,
+though it may in the future.
+
+Currently, it can do the following:
+
+- Create a header with any name
+- Populate that header with any credential
+- Populate that header any combination of credentials and literal strings
+- Populate that header any supported transformation of any combination of
+  credentials and fixed strings
+
+Basically, you are limited only by what's possible in the Go `text/template`
+package, and by the additional functions Secretless makes available.
+
+Currently, the only additonal function beyond the defaults of `text/template`
+is `base64`, but we plan to add more and can do so easily.  If there's one
+you'd like to see, please create an issue or pull request.
+
+## Developer Documentation
+
+For developers, the `generic` package makes writing new HTTP connectors easy.
+You can create a new connector with just a bit of boilerplate and a single
+declarative struct to define the Authorization headers.
+
+### Defining your connector
+
+You'll define your generic connector using the `ConfigYAML` type.  Essentially,
+`ConfigYAML` lets you use Go code to write the same configuration that
+end-users write in `secretless.yml`.  
+
+Let's take a look at the Basic auth connector to see how this works:
+
+```go
+// Taken from:
+// internal/plugin/connectors/http/basicauth/plugin.go
+//
+// Also note: 
+// "NewConnectorConstructor" and "ConfigYAML" are defined in:
+// /internal/plugin/connectors/http/generic/external_api.go
+
+newConnector, err := generichttp.NewConnectorConstructor(
+  &generichttp.ConfigYAML{
+    CredentialValidations: map[string]string{
+      "username": "[^:]+",
+    },
+    Headers: map[string]string{
+      "Authorization": "Basic {{ printf \"%s:%s\" .username .password | base64 }}",
+    },
+  },
+)
+``` 
+
+Note how this code exactly mirrors the yaml from the end-user example.  Please
+refer to those docs above for an explanation of the `CredentialValidations` and
+`Headers` formats.
+
+`ConfigYAML`, then, defines your connector, and `NewConnectorConstructor` is a
+convenience function to transform that definition into an `http.Plugin`.
+
+### Adding your connector
+
+While the heart of creating a connector is writing the `ConfigYAML` definition 
+described above, a few other pieces of boilerplate are required to complete
+the process.
+
+Refer to the Basic auth connector
+(`internal/plugin/connectors/http/basicauth/plugin.go`) as an example when
+working through the steps below.
+
+Here are the steps, in detail:
+
+1. Create a new directory for your connector under `/internal/plugin/connectors/http`.
```

Comment:
Not a blocker: I feel like we've cut away so much code already with the success of the generic connector, perhaps we can reduce this to a single connector directory containing all the various plugin files, with appropriate naming conventions for each file. 

There's a part of me that wonders if there's value in reducing it further, since the reduced plugin files are so similar now. However, it might reduce readability.

Reply:
Interesting.  So the idea would be to have a directory like `/internal/plugin/connectors/http/generic_implementations` and then files `basic_auth.go` and `conjur.go`, etc in there?

I sort of like the idea but I'm not sure the incremental benefit makes it worth doing.  It also does have a couple downsides: eg, the test dirs would no longer mirror the impl dirs.  Also, perhaps more importantly, you're breaking the consistency of things, ie, making the generic connectors into a special case.  Whereas currently they're just like any other connector and they _happen_ to be implemented using some utilities provided by the `generic` classes.  That might be a preferable way to think of them.  It's a debatable point.

In any case, if we decide to do this, it will involve changes beyond this README so we should propose it as a followup.